src/client/mapblock_mesh.cpp

```diff
@@ -1003,6 +1004,176 @@ static void applyTileColor(PreMeshBuffer &pmb)
 	}
 }
 
+/*
+	MapBlockBspTree
+*/
+
+void MapBlockBspTree::buildTree(const std::vector<MeshTriangle> *triangles)
+{
+	this->triangles = triangles;
+
+	nodes.clear();
+
+	std::vector<s32> indexes;
+	for (u16 i = 0; i < triangles->size(); i++)
+		indexes.push_back(i);
+
+	root = buildTree(v3f(1, 0, 0), v3f(85, 85, 85), 40, indexes, 0);
+}
+
+static const MeshTriangle *findSplitCandidate(const std::vector<s32> &list, const std::vector<MeshTriangle> &triangles)
+{
+	// find best origin as a center of the cluster.
+	v3f origin(0, 0, 0);
+	size_t n = list.size();
+	for (s32 i : list) {
+		origin += triangles[i].centroid / n;
+	}
+
+	// find the triangle with the largest area and closest to the center
+	const MeshTriangle *candidate_triangle = &triangles[list[0]];
+	const MeshTriangle *ith_triangle;
+	for (s32 i : list) {
+		ith_triangle = &triangles[i];
+		if (ith_triangle->areaSQ > candidate_triangle->areaSQ ||
+				(ith_triangle->areaSQ == candidate_triangle->areaSQ &&
+				ith_triangle->centroid.getDistanceFromSQ(origin) < candidate_triangle->centroid.getDistanceFromSQ(origin))) {
+			candidate_triangle = ith_triangle;
+		}
+	}
+	return candidate_triangle;
+}
+
+s32 MapBlockBspTree::buildTree(v3f normal, v3f origin, float delta, const std::vector<s32> &list, u32 depth)
+{
+	// if the list is empty, don't bother
+	if (list.size() == 0) {
```

Comment:
generally you can replace all `.size() == 0` or `.size() > 0` here with `.empty()` or `!.empty()` respectively

Reply:
replaced